
## Assignment 3 (Regularization) by vipin_2011MT22

Implement L2, L1, early stopping and drop out regularization for the previous assignment and write a report comparing the results (Accuracy, Precision, Recall and F1). 

In [ ]:
#importing the libraries
import pandas as pd
import numpy as np
from keras.models import Sequential
from keras.layers import Dense, Dropout
from keras.callbacks import EarlyStopping
from sklearn.model_selection import train_test_split
from keras.callbacks import ModelCheckpoint
from keras.constraints import maxnorm
from tensorflow.keras import regularizers
from sklearn.metrics import precision_score, recall_score, accuracy_score, f1_score

In [ ]:
#importing the dataset
dataframe = pd.read_csv("diabetes.csv")
dataframe.head()

Pregnancies  Glucose  BloodPressure  ...  DiabetesPedigreeFunction  Age  Outcome
0            6      148             72  ...                     0.627   50        1
1            1       85             66  ...                     0.351   31        0
2            8      183             64  ...                     0.672   32        1
3            1       89             66  ...                     0.167   21        0
4            0      137             40  ...                     2.288   33        1

[5 rows x 9 columns]

In [ ]:
#spilting of dataset into feature and label
df_label = dataframe['Outcome']
df_features = dataframe.drop('Outcome', 1)
df_features.replace('?', -99999, inplace=True)
print(df_label.head())
print(df_features.head())

0    1
1    0
2    1
3    0
4    1
Name: Outcome, dtype: int64
   Pregnancies  Glucose  BloodPressure  ...   BMI  DiabetesPedigreeFunction  Age
0            6      148             72  ...  33.6                     0.627   50
1            1       85             66  ...  26.6                     0.351   31
2            8      183             64  ...  23.3                     0.672   32
3            1       89             66  ...  28.1                     0.167   21
4            0      137             40  ...  43.1                     2.288   33

[5 rows x 8 columns]


In [ ]:
#hot encoding the label dataset
label = []
for lab in df_label:
    if lab == 1:
        label.append([1, 0])  # class 1
    elif lab == 0:
        label.append([0, 1])  # class 0
      

In [ ]:
data = np.array(df_features)
label = np.array(label)
print(data.shape,label.shape)

(768, 8) (768, 2)


In [ ]:
#spilting dataset into testing and training
x_train, x_test, y_train, y_test = train_test_split(data, label, test_size=0.2, random_state=2)
x_train.shape

(614, 8)

# For l2

In [ ]:
#building our Neural Network
model = Sequential()
model.add(Dense(500, input_dim=8, activation='sigmoid',kernel_regularizer=regularizers.l2(0.01), bias_regularizer=regularizers.l2(0.01), 
                activity_regularizer=regularizers.l2(0.01)))
model.add(Dense(100, activation='sigmoid',kernel_regularizer=regularizers.l2(0.01), bias_regularizer=regularizers.l2(0.01), 
                activity_regularizer=regularizers.l2(0.01)))
model.add(Dense(2, activation='softmax'))
model.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])
model.fit(x_train,y_train, epochs=1000, batch_size=70, validation_data=(x_test, y_test))

Epoch 1/1000
9/9 [==============================] - 1s 62ms/step - loss: 4.9969 - accuracy: 0.5040 - val_loss: 4.1902 - val_accuracy: 0.7078
Epoch 2/1000
9/9 [==============================] - 0s 9ms/step - loss: 4.0711 - accuracy: 0.6400 - val_loss: 3.4837 - val_accuracy: 0.7078
Epoch 3/1000
9/9 [==============================] - 0s 7ms/step - loss: 3.3932 - accuracy: 0.6151 - val_loss: 2.8637 - val_accuracy: 0.7078
Epoch 4/1000
9/9 [==============================] - 0s 7ms/step - loss: 2.7929 - accuracy: 0.6133 - val_loss: 2.3187 - val_accuracy: 0.7078
Epoch 5/1000
9/9 [==============================] - 0s 7ms/step - loss: 2.2719 - accuracy: 0.6292 - val_loss: 1.8990 - val_accuracy: 0.7078
Epoch 6/1000
9/9 [==============================] - 0s 7ms/step - loss: 1.8678 - accuracy: 0.6500 - val_loss: 1.5995 - val_accuracy: 0.7078
Epoch 7/1000
9/9 [==============================] - 0s 9ms/step - loss: 1.5873 - accuracy: 0.6369 - val_loss: 1.4041 - val_accuracy: 0.7078
Epoch 8/1000
9/9 [=

In [ ]:
model.summary()

Model: "sequential"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense (Dense)                (None, 500)               4500      
_________________________________________________________________
dense_1 (Dense)              (None, 100)               50100     
_________________________________________________________________
dense_2 (Dense)              (None, 2)                 202       
Total params: 54,802
Trainable params: 54,802
Non-trainable params: 0
_________________________________________________________________


In [ ]:
#predictating whether diabetic or not
Y_pred = model.predict(x_test)
Y_pred =(Y_pred>0.5)

In [ ]:
#Calculating accuracy, f1 score, precision, recall

accuracyL2 = accuracy_score(y_test, Y_pred)
scoreL2 = f1_score(y_test, Y_pred, average='weighted')
precisionL2 = precision_score(y_test, Y_pred, labels=[0,1], average='weighted')
recallL2 = recall_score(y_test, Y_pred, average='weighted')

print('accuracy: ',accuracyL2)
print('F1-score: ',scoreL2)
print('Precision: ', precisionL2)
print("recall: ", recallL2)



accuracy:  0.7662337662337663
F1-score:  0.7589679907832293
Precision:  0.7563457563457563
recall:  0.7662337662337663


# For l1

In [ ]:
#building our Neural Network
model = Sequential()
model.add(Dense(500, input_dim=8, activation='sigmoid',kernel_regularizer=regularizers.l1(0.0001), bias_regularizer=regularizers.l1(0.0001), 
                activity_regularizer=regularizers.l1(0.0001)))
model.add(Dense(100, activation='sigmoid',kernel_regularizer=regularizers.l1(0.0001), bias_regularizer=regularizers.l1(0.0001), 
                activity_regularizer=regularizers.l1(0.0001)))
model.add(Dense(2, activation='softmax'))
model.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])
model.fit(x_train,y_train, epochs=1000, batch_size=70, validation_data=(x_test, y_test))

Epoch 1/1000
9/9 [==============================] - 1s 29ms/step - loss: 1.0600 - accuracy: 0.5612 - val_loss: 0.9176 - val_accuracy: 0.7078
Epoch 2/1000
9/9 [==============================] - 0s 8ms/step - loss: 0.9787 - accuracy: 0.6179 - val_loss: 0.8942 - val_accuracy: 0.6948
Epoch 3/1000
9/9 [==============================] - 0s 8ms/step - loss: 0.9042 - accuracy: 0.6539 - val_loss: 0.8453 - val_accuracy: 0.7078
Epoch 4/1000
9/9 [==============================] - 0s 8ms/step - loss: 0.8635 - accuracy: 0.6711 - val_loss: 0.8345 - val_accuracy: 0.7338
Epoch 5/1000
9/9 [==============================] - 0s 8ms/step - loss: 0.8565 - accuracy: 0.6731 - val_loss: 0.8156 - val_accuracy: 0.7338
Epoch 6/1000
9/9 [==============================] - 0s 7ms/step - loss: 0.8343 - accuracy: 0.6860 - val_loss: 0.7975 - val_accuracy: 0.7468
Epoch 7/1000
9/9 [==============================] - 0s 8ms/step - loss: 0.8215 - accuracy: 0.6943 - val_loss: 0.7809 - val_accuracy: 0.7273
Epoch 8/1000
9/9 [=

In [ ]:
model.summary()

Model: "sequential_1"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_3 (Dense)              (None, 500)               4500      
_________________________________________________________________
dense_4 (Dense)              (None, 100)               50100     
_________________________________________________________________
dense_5 (Dense)              (None, 2)                 202       
Total params: 54,802
Trainable params: 54,802
Non-trainable params: 0
_________________________________________________________________


In [ ]:
#predictating whether diabetic or not
Y_pred = model.predict(x_test)
Y_pred =(Y_pred>0.5)

In [ ]:
#Calculating accuracy, f1 score, precision, recall

accuracyL1 = accuracy_score(y_test, Y_pred)
scoreL1 = f1_score(y_test, Y_pred, average='weighted')
precisionL1 = precision_score(y_test, Y_pred, labels=[0,1], average='weighted')
recallL1 = recall_score(y_test, Y_pred, average='weighted')

print('accuracy: ',accuracyL1)
print('F1-score: ',scoreL1)
print('Precision: ', precisionL1)
print("recall: ", recallL1)

accuracy:  0.7077922077922078
F1-score:  0.708724455268234
Precision:  0.7097162097162096
recall:  0.7077922077922078


# For Early Stopping

In [ ]:
#building our Neural Network
model = Sequential()
model.add(Dense(500, input_dim=8, activation='sigmoid'))
model.add(Dense(100, activation='sigmoid'))
model.add(Dense(2, activation='softmax'))
model.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])

In [ ]:
es = EarlyStopping(monitor='val_loss',patience=300)
mc = ModelCheckpoint('best_model.h5',monitor='val_accuracy',mode='max',verbose=0,save_best_only=True)
model.fit(x_train,y_train, epochs=1000, batch_size=70, validation_data=(x_test,y_test),callbacks=[es, mc])

Epoch 1/1000
9/9 [==============================] - 1s 24ms/step - loss: 0.7462 - accuracy: 0.5189 - val_loss: 0.6020 - val_accuracy: 0.7078
Epoch 2/1000
9/9 [==============================] - 0s 7ms/step - loss: 0.6839 - accuracy: 0.6161 - val_loss: 0.6260 - val_accuracy: 0.6818
Epoch 3/1000
9/9 [==============================] - 0s 7ms/step - loss: 0.6281 - accuracy: 0.6603 - val_loss: 0.5721 - val_accuracy: 0.7078
Epoch 4/1000
9/9 [==============================] - 0s 7ms/step - loss: 0.6103 - accuracy: 0.6426 - val_loss: 0.5846 - val_accuracy: 0.7273
Epoch 5/1000
9/9 [==============================] - 0s 7ms/step - loss: 0.6062 - accuracy: 0.6661 - val_loss: 0.5552 - val_accuracy: 0.7468
Epoch 6/1000
9/9 [==============================] - 0s 7ms/step - loss: 0.5921 - accuracy: 0.6919 - val_loss: 0.5605 - val_accuracy: 0.7403
Epoch 7/1000
9/9 [==============================] - 0s 8ms/step - loss: 0.5595 - accuracy: 0.7143 - val_loss: 0.5524 - val_accuracy: 0.7468
Epoch 8/1000
9/9 [=

In [ ]:
model.summary()

Model: "sequential_8"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_24 (Dense)             (None, 500)               4500      
_________________________________________________________________
dense_25 (Dense)             (None, 100)               50100     
_________________________________________________________________
dense_26 (Dense)             (None, 2)                 202       
Total params: 54,802
Trainable params: 54,802
Non-trainable params: 0
_________________________________________________________________


In [ ]:
#predictating whether diabetic or not
Y_pred = model.predict(x_test)
Y_pred =(Y_pred>0.5)

In [ ]:
#Calculating accuracy, f1 score, precision, recall

accuracyES = accuracy_score(y_test, Y_pred)
scoreES = f1_score(y_test, Y_pred, average='weighted')
precisionES = precision_score(y_test, Y_pred, labels=[0,1], average='weighted')
recallES = recall_score(y_test, Y_pred, average='weighted')

print('accuracy: ',accuracyES)
print('F1-score: ',scoreES)
print('Precision: ', precisionES)
print("recall: ", recallES)



accuracy:  0.7402597402597403
F1-score:  0.743358649701859
Precision:  0.7474688576729392
recall:  0.7402597402597403


# For Dropout

In [ ]:
#building our Neural Network
model = Sequential()
model.add(Dense(500, input_dim=8, activation='sigmoid',kernel_constraint=maxnorm(3)))
model.add(Dropout(0.3))

model.add(Dense(100, activation='sigmoid',kernel_constraint=maxnorm(3)))
model.add(Dropout(0.3))

model.add(Dense(2, activation='softmax'))
model.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])
model.fit(x_train,y_train, epochs=1000, batch_size=70, validation_data=(x_test, y_test))

Epoch 1/1000
9/9 [==============================] - 1s 23ms/step - loss: 0.6998 - accuracy: 0.5995 - val_loss: 0.5902 - val_accuracy: 0.7078
Epoch 2/1000
9/9 [==============================] - 0s 7ms/step - loss: 0.6881 - accuracy: 0.6176 - val_loss: 0.5792 - val_accuracy: 0.7078
Epoch 3/1000
9/9 [==============================] - 0s 8ms/step - loss: 0.6518 - accuracy: 0.6068 - val_loss: 0.5746 - val_accuracy: 0.7143
Epoch 4/1000
9/9 [==============================] - 0s 8ms/step - loss: 0.6517 - accuracy: 0.6242 - val_loss: 0.5681 - val_accuracy: 0.7273
Epoch 5/1000
9/9 [==============================] - 0s 9ms/step - loss: 0.6012 - accuracy: 0.6751 - val_loss: 0.5571 - val_accuracy: 0.7468
Epoch 6/1000
9/9 [==============================] - 0s 9ms/step - loss: 0.6231 - accuracy: 0.6482 - val_loss: 0.5566 - val_accuracy: 0.7468
Epoch 7/1000
9/9 [==============================] - 0s 7ms/step - loss: 0.6039 - accuracy: 0.6847 - val_loss: 0.5504 - val_accuracy: 0.7403
Epoch 8/1000
9/9 [=

In [ ]:
model.summary()

Model: "sequential_3"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_9 (Dense)              (None, 500)               4500      
_________________________________________________________________
dropout (Dropout)            (None, 500)               0         
_________________________________________________________________
dense_10 (Dense)             (None, 100)               50100     
_________________________________________________________________
dropout_1 (Dropout)          (None, 100)               0         
_________________________________________________________________
dense_11 (Dense)             (None, 2)                 202       
Total params: 54,802
Trainable params: 54,802
Non-trainable params: 0
_________________________________________________________________


In [ ]:
#predictating whether diabetic or not
Y_pred = model.predict(x_test)
Y_pred =(Y_pred>0.5)

In [ ]:
#Calculating accuracy, f1 score, precision, recall

accuracyDO = accuracy_score(y_test, Y_pred)
scoreDO = f1_score(y_test, Y_pred, average='weighted')
precisionDO = precision_score(y_test, Y_pred, labels=[0,1], average='weighted')
recallDO = recall_score(y_test, Y_pred, average='weighted')

print('accuracy: ',accuracyDO)
print('F1-score: ',scoreDO)
print('Precision: ', precisionDO)
print("recall: ", recallDO)

accuracy:  0.7077922077922078
F1-score:  0.7178683385579938
Precision:  0.7409505388228792
recall:  0.7077922077922078


#Report

In [ ]:
print('For L2')
print('accuracy: ',accuracyL2)
print('F1-score: ',scoreL2)
print('Precision: ', precisionL2)
print("recall: ", recallL2, '\n')

print('For L1')
print('accuracy: ',accuracyL1)
print('F1-score: ',scoreL1)
print('Precision: ', precisionL1)
print("recall: ", recallL1, '\n')

print('For Early Stopping')
print('accuracy: ',accuracyES)
print('F1-score: ',scoreES)
print('Precision: ', precisionES)
print("recall: ", recallES, '\n')

print('For Dropout')
print('accuracy: ',accuracyDO)
print('F1-score: ',scoreDO)
print('Precision: ', precisionDO)
print("recall: ", recallDO, '\n')

For L2
accuracy:  0.7662337662337663
F1-score:  0.7589679907832293
Precision:  0.7563457563457563
recall:  0.7662337662337663 

For L1
accuracy:  0.7077922077922078
F1-score:  0.708724455268234
Precision:  0.7097162097162096
recall:  0.7077922077922078 

For Early Stopping
accuracy:  0.7402597402597403
F1-score:  0.724173285879275
Precision:  0.7220922272063107
recall:  0.7402597402597403 

For Dropout
accuracy:  0.7077922077922078
F1-score:  0.7178683385579938
Precision:  0.7409505388228792
recall:  0.7077922077922078 



L2 regularization model is giving maximum accuracy followed by EarlyStopping. Dropout and L1 regularization model is giving same accuracy.

F1-score of  L2 regularization model is maximum followed by Dropout regularization model, EarlyStopping regularization model and L1 regularization model.

Pricision of L2 regularization model is maximum followed by Dropout regularization model, L1 regularization model and EarlyStopping regularization model.

L2 regularization model is giving maximum recall followed by EarlyStopping regularization model, Dropout regularization model and L1 regularization model 

for early stopping, the patience parameter plays a huge role, for patience = 300, model stopped at 373 epochs with accuracy of around 74% on test data. This gives a hint that epochs = 1000 is bound to give an overfitted model

comparing between L1 and L2 when for the same models we used a parameter of 0.01 for both the regularisations, L2 had better results and L1 seemed overfitted

overall the metrics are comparable for all models however the L2 model seems to perform slightly better than other models